### PR TITLE
Implement endpoint watch and resolver

### PR DESCRIPTION
**Pull request by lemonlinger**
Fix https://github.com/etcd-io/etcd/issues/12652

@peterbourgon

### CODE REVIEW

File: client/v3/naming/endpoints/endpoints_impl.go
Comment (ptabor) [inline client/v3/naming/endpoints/endpoints_impl.go:89]: Maybe worth logging the error at least (as WARNING with the 'corrupted' key)

Comment (lemonlinger) [reply client/v3/naming/endpoints/endpoints_impl.go:89]: Should the logger(`*zap.Logger`) be created internally or passed from outside?

Comment (ptabor) [reply client/v3/naming/endpoints/endpoints_impl.go:89]: Good question. 

1. I assumed clientv3.lg, but this one is private. 
2. logger.go is practically dead  - I will prepare PR to delete this one. 

So there are 2 options: 
  1. Expose the public c.Logger() getter and use it. 
  2. Expand signature of NewManager to take the logger parameter. 
  
@gyuho @jingyih @jpbetz -> do you have opinion here ?

File: client/v3/naming/resolver/resolver.go
Comment (ptabor) [inline client/v3/naming/resolver/resolver.go:100]: Please add comment: 
 // It's just a hint, resolver can ignore this if it's not necessary.

File: tests/integration/clientv3/naming/resolver_test.go
Comment (ptabor) [inline tests/integration/clientv3/naming/resolver_test.go:73]: I'm curious ? Do we need tripple: etcd:///foo or 'etcd://foo' is sufficient

Please update: 
https://github.com/etcd-io/website/blob/master/content/docs/current/dev-guide/grpc_naming.md if we need '///'

**Review by ptabor — APPROVED** [review]
Thank you. That looks really good. 

Let's wait for other maintainers feedback/suggestions e.g. ad. logger.

File: tests/integration/clientv3/naming/resolver_test.go
Comment (lemonlinger) [reply tests/integration/clientv3/naming/resolver_test.go:73]: Yes. According to this official doc: https://github.com/grpc/grpc/blob/master/doc/naming.md#name-syntax, gRPC cannot parse `etcd://foo` properly to get the expected scheme `etcd`.

File: client/v3/naming/endpoints/endpoints_impl.go
Comment (ptabor) [reply client/v3/naming/endpoints/endpoints_impl.go:89]: OK. Let's go for 1.  It's in scope of the same 'library' and it's hard to imagine use-case when user want's to have separate logger for endpoint resolution. 

Please comment that this method is for internal use of etcd-client library and should not be used as general-purpose logger.

File: tests/integration/clientv3/naming/resolver_test.go
Comment (ptabor) [inline tests/integration/clientv3/naming/resolver_test.go:111]: FYI: 

git commit -amend 
git push --force

Is usual "hack" to re-trigger tests.

Comment (lemonlinger) [reply tests/integration/clientv3/naming/resolver_test.go:111]: Thanks!

File: client/v3/naming/endpoints/endpoints_impl.go
Comment (lemonlinger) [reply client/v3/naming/endpoints/endpoints_impl.go:89]: OK

**Review by ptabor — APPROVED** [review]

**Comment by ptabor** [comment]
Thank you

File: client/v3/client.go
Comment (ahrtr) [inline client/v3/client.go:116]: @chaochn47 It seems that you missed this change in the second commit in https://github.com/etcd-io/etcd/pull/16800?

Comment (chaochn47) [reply client/v3/client.go:116]: Thanks for capturing that! 

Zero value of `RWMutex` should be the same `new(sync.RWMutext)` though.

Comment (ahrtr) [reply client/v3/client.go:116]: > Zero value of RWMutex should be the same new(sync.RWMutext) though.


It isn't true. zero value for a pointer is nil,  while the result of `new(sync.RWMutext)` isn't nil.

Comment (chaochn47) [reply client/v3/client.go:116]: You are right.

**Pull request merged**
